Course/Course/Presentation/Dates/CourseDatesViewModel.swift

```diff
@@ -91,4 +93,52 @@ public class CourseDatesViewModel: ObservableObject {
             errorMessage = CourseLocalization.Error.componentNotFount
         }
     }
+    
+    @MainActor
+    func shiftDueDates(courseID: String, withProgress: Bool = true) async {
+        isShowProgress = withProgress
+        dueDatesShifted = false
+        do {
+            try await interactor.shiftDueDates(courseID: courseID)
+            NotificationCenter.default.post(name: .shiftCourseDates, object: courseID)
+            isShowProgress = false
+        } catch let error {
+            isShowProgress = false
+            if error.isInternetError || error is NoCachedDataError {
+                errorMessage = CoreLocalization.Error.slowOrNoInternetConnection
+            } else {
+                errorMessage = CoreLocalization.Error.unknownError
+            }
+        }
+    }
+    
+    deinit {
+        NotificationCenter.default.removeObserver(self)
+    }
+}
+
+extension CourseDatesViewModel {
+    private func addObservers() {
+        NotificationCenter.default.addObserver(
+            self,
+            selector: #selector(handleShiftDueDates(_:)),
+            name: .shiftCourseDates, object: nil
+        )
+    }
+    
+    @objc private func handleShiftDueDates(_ notification: Notification) {
+        if let courseID = notification.object as? String {
+            Task {
+                await self.getCourseDates(courseID: courseID)
+                DispatchQueue.main.async { [weak self] in
+                    guard let self else { return }
+                    self.dueDatesShifted = true
```

Comment:
Do we need to check wether the `self` exists or not, can't `self?.dueDatesShifted = true` be ignored if the self will be nil?

Reply:
Same here, please do not combine Swift Concurrency and GCD frameworks.
Use the Main Actor instead.